plugins/catalog/src/components/EntitySwitch/conditions.ts

```diff
@@ -37,34 +38,48 @@ function strCmpAll(value: string | undefined, cmpValues: string | string[]) {
  * @public
  */
 export function isKind(kinds: string | string[]) {
-  return (entity: Entity) => strCmpAll(entity.kind, kinds);
+  return isEntityWith({ kind: kinds });
 }
 
 /**
  * For use in EntitySwitch.Case. Matches if the entity is a Component of a given spec.type.
  * @public
  */
 export function isComponentType(types: string | string[]) {
-  return (entity: Entity) => {
-    if (!strCmp(entity.kind, 'component')) {
-      return false;
-    }
-    const componentEntity = entity as ComponentEntity;
-    return strCmpAll(componentEntity.spec.type, types);
-  };
+  return isEntityWith({ kind: 'component', type: types });
 }
 
 /**
  * For use in EntitySwitch.Case. Matches if the entity is a Resource of a given spec.type.
  * @public
  */
 export function isResourceType(types: string | string[]) {
+  return isEntityWith({ kind: 'resource', type: types });
+}
+
+/**
+ *
+ * For use in EntitySwitch.Case. Matches if the entity is the specified kind and type (if present).
+ * @public
+ */
+export function isEntityWith(predicate: EntityPredicates) {
   return (entity: Entity) => {
-    if (!strCmp(entity.kind, 'resource')) {
+    if (!strCmpAll(entity.kind, predicate.kind)) {
       return false;
     }
-    const resourceEntity = entity as ResourceEntity;
-    return strCmpAll(resourceEntity.spec.type, types);
+
+    if (
+      !predicate.type ||
+      (Array.isArray(predicate.type) && predicate.type.length === 0)
```

Comment:
Should probably do

```
const types = [predicate.type].flat().filter(p => typeof p === 'string');
if (types.length && !strCmpAll(entity.spec.type, types)) {
  return false;  
}

return true;
```

Come to think of it, I think we might want to improve strCmpAll while we're at it. We want its first argument to be of type `unknown`, so that we can for example do `strCmpAll((entity as any).spec?.type, ...` without knowing that the actual entity even has a type field, much less even a spec. Then `strCmpAll` needs to check that the type of the incoming argument really is a string, so it fails if it's anything else

Reply:
Good call. I pushed the `unknown` (and corresponding type checks) down into `strCmp`.